// runflick/dist/plugin.js
"use strict";
// Plugin system for Flick
Object.defineProperty(exports, "__esModule", { value: true });
exports.PluginManager = void 0;
class PluginManager {
    plugins = new Map();
    declaredPlugins = new Map();
    registerPlugin(plugin) {
        this.plugins.set(plugin.name, plugin);
    }
    getPlugin(name) {
        return this.plugins.get(name);
    }
    declarePlugin(name, args = null) {
        const plugin = this.plugins.get(name);
        if (!plugin) {
            throw new Error(`Unknown plugin: ${name}`);
        }
        this.declaredPlugins.set(name, args);
        if (plugin.onDeclare) {
            plugin.onDeclare(args);
        }
    }
    isDeclared(name) {
        return this.declaredPlugins.has(name);
    }
    getPluginArgs(name) {
        return this.declaredPlugins.get(name);
    }
    getDeclaredPlugins() {
        return this.declaredPlugins;
    }
    async runFileCompleteHooks(context) {
        for (const [name, args] of this.declaredPlugins) {
            const plugin = this.plugins.get(name);
            if (plugin?.onFileComplete) {
                await plugin.onFileComplete(context);
            }
        }
    }
    reset() {
        this.declaredPlugins.clear();
    }
}
exports.PluginManager = PluginManager;
